proc-macro-helpers/src/lib.rs

```diff
@@ -9,15 +9,18 @@
 
 extern crate frunk_core;
 extern crate proc_macro;
+extern crate proc_macro2;
 
 #[macro_use]
 extern crate quote;
 extern crate syn;
 
 use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
```

Comment:
Interesting; just for my own curiosity, is there any reason why we wouldn't pick one of either `proc_macro` or `proc_macro2`?

Reply:
Honestly I'm not super familiar with all the details of how procedural macros work nowadays. The course of events was:
- I had to replace the use of `impl ToTokens` with a concrete type due to this issue with `impl Trait`: https://old.reddit.com/r/rust/comments/c2ttxg/send_and_sync_leak_through_impl_trait_because/ermiyyu/

- The concrete type returned by the `quote!` macro is from `proc_macro2`, so I had to import that.

My limited understanding is that `proc_macro::TokenStream` is tied to the compiler's plugin architecture (ie. you *can't* use `proc_macro2` in the top-level signature), and `proc_macro2::TokenStream` is used by `quote`, etc. so that the APIs can evolve separately (avoiding the issue where compiler releases would break `syn` frequently) and so both are necessary.

Reply:
To my knowledge, it is very common for a crate to need both `proc_macro` and `proc_macro2` TokenStreams. (`proc_macro` for the entry points, `proc_macro2` for everything else).

* https://github.com/dtolnay/indoc/blob/master/impl/src/lib.rs
* https://github.com/serde-rs/serde/blob/master/serde_derive/src/de.rs

Reply:
Wow, TIL (2 things actually, that there was an **old**.reddit.com, and this leakage).  Somehow the leakage seems .... desirable to me, but I haven't really given it more than a passing thought.

> * I had to replace the use of `impl ToTokens` with a concrete type due to this issue with `impl Trait`: https://old.reddit.com/r/rust/comments/c2ttxg/send_and_sync_leak_through_impl_trait_because/ermiyyu/

Again, for my own learning: was using `impl ToTokens` as a return type causing us problems?

Reply:
Yes, due to the implicit lifetimes, if you replace all instances of `-> TokenStream2` with `-> impl ToToken` and try to compile, you will get several errors like the following:

```
error: cannot infer an appropriate lifetime
   --> proc-macro-helpers\src\lib.rs:273:38
    |
272 |     pub fn build_hlist_type<R: ToTokens>(&self, f: impl Fn(&FieldBinding) -> R) -> impl ToTokens {
    |                                                                                    ------------- this return type evaluates to the `'static` lifetime...
273 |         build_hlist_type(self.fields.iter().map(f))
    |                          ----------- ^^^^
    |                          |
    |                          ...but this borrow...
    |
note: ...can't outlive the anonymous lifetime #1 defined on the method body at 272:5
   --> proc-macro-helpers\src\lib.rs:272:5
    |
272 | /     pub fn build_hlist_type<R: ToTokens>(&self, f: impl Fn(&FieldBinding) -> R) -> impl ToTokens {
273 | |         build_hlist_type(self.fields.iter().map(f))
274 | |     }
    | |_____^
help: you can add a constraint to the return type to make it last less than `'static` and match the anonymous lifetime #1 defined on the method body at 272:5
    |
272 |     pub fn build_hlist_type<R: ToTokens>(&self, f: impl Fn(&FieldBinding) -> R) -> impl ToTokens + '_ {
    |                                                                                    ^^^^^^^^^^^^^^^^^^
```

Reply:
Adding the `'_` lifetime as suggested does not help: the problem here is that the lifetime of the return type is *always* `'static` regardless of the lifetime on `&self`, because `TokenStream2` has no lifetime specifiers, but there's no way to make `impl Trait` understand that.

Reply:
Aahhh, understood. Thanks for the answers I’ve learnt a lot :)